features/netsocket/TCPSocket.cpp

```diff
@@ -39,10 +39,47 @@ nsapi_error_t TCPSocket::connect(const SocketAddress &address)
     _lock.lock();
     nsapi_error_t ret;
 
-    if (!_socket) {
-        ret = NSAPI_ERROR_NO_SOCKET;
-    } else {
+    // If this assert is hit then there are two threads
+    // performing a send at the same time which is undefined
+    // behavior
+    MBED_ASSERT(!_write_in_progress);
+    _write_in_progress = true;
+
+    bool blocking_connect_in_progress = false;
+
+    while (true) {
+        if (!_socket) {
+            ret = NSAPI_ERROR_NO_SOCKET;
+            break;
+        }
+
+        _pending = 0;
         ret = _stack->socket_connect(_socket, address);
+        if ((_timeout == 0) || !(ret == NSAPI_ERROR_IN_PROGRESS || ret == NSAPI_ERROR_ALREADY)) {
+            break;
+        } else {
+            blocking_connect_in_progress = true;
+
+            int32_t count;
+
+            // Release lock before blocking so other threads
+            // accessing this object aren't blocked
+            _lock.unlock();
+            count = _write_sem.wait(_timeout);
+            _lock.lock();
+
+            if (count < 1) {
+                // Semaphore wait timed out so break out and return
+                break;
+            }
+        }
+    }
+
+    _write_in_progress = false;
+
+    /* Non-blocking connect gives "EISCONN" once done - convert to OK for blocking mode if we became connected during this call */
+    if (ret == NSAPI_ERROR_IS_CONNECTED && blocking_connect_in_progress) {
```

Comment:
Should this check just be `ret == NSAPI_ERROR_IS_CONNECTED && _timeout > 0`? What if a socket returns NSAPI_ERROR_IS_CONNECTED on the first call?